### ISSUE TITLE
git add --interactive broken

### ISSUE THREAD

**Issue by axeld** [haikuports/haikuports#2300]
When I try to use "git add -i" the following message appears:
`Can't locate Git.pm in @INC (you may need to install the Git module) (@INC contains: /packages/git-2.16.2-1/.self/non-packaged/lib/perl5/site_perl/5.26 /boot/home/config/lib/perl5/5.26.1/BePC-haiku /boot/home/config/lib/perl5/5.26.1 /boot/home/config/non-packaged/lib/perl5/site_perl/5.26.1/BePC-haiku /boot/home/config/non-packaged/lib/perl5/site_perl/5.26.1 /boot/home/config/lib/perl5/vendor_perl/5.26.1/BePC-haiku /boot/home/config/lib/perl5/vendor_perl/5.26.1 /boot/home/config/lib/perl5/vendor_perl /boot/system/lib/perl5/5.26.1/BePC-haiku /boot/system/lib/perl5/5.26.1 /boot/system/non-packaged/lib/perl5/site_perl/5.26.1/BePC-haiku /boot/system/non-packaged/lib/perl5/site_perl/5.26.1 /boot/system/lib/perl5/vendor_perl/5.26.1/BePC-haiku /boot/system/lib/perl5/vendor_perl/5.26.1 /boot/system/lib/perl5/vendor_perl) at /packages/git-2.16.2-1/.self/lib/git-core/git-add--interactive line 7.
BEGIN failed--compilation aborted at /packages/git-2.16.2-1/.self/lib/git-core/git-add--interactive line 7.`

I'm using the package "git-2.16.2-1". This still worked until recently.

**Comment by fbrosson** [comment]
I have the same problem on x86_gcc2 and x86_64.
On x86, however, I have git 2.16.2 and perl 5.18.2 and `git add -i` is working fine.
```
~> git version
git version 2.16.2
~> perl -v | head -5

This is perl 5, version 18, subversion 2 (v5.18.2) built for BePC-haiku
(with 1 registered patch, see perl -V for more detail)

Copyright 1987-2013, Larry Wall

~> git add -i

*** Commands ***
  1: status	  2: update	  3: revert	  4: add untracked
  5: patch	  6: diff	  7: quit	  8: help
What now> 
```

**Comment by fbrosson** [comment]
I have built git-2.15.1 on x86_64 (with perl 5.26.1) and I get the same error with `git add -i`.
I'm building perl-5.18.2-3 and will try to rebuild git to see if this fixes the issue.

I'm not telling we should downgrade to perl 5.18.2. I'm just trying to find where the problem is. Maybe git needs some fixes to work with perl 5.26.1.
I'll let the perl experts decide whether or not a temporary downgrade is appropriate.

**Comment by fbrosson** [comment]
Build and installtion OK:
```
The following changes will be made:
  in system:
    upgrade package perl-5.26.1-1 to 5.18.2-3 from local file
    upgrade package git-2.15.1-2 to 2.16.2-1 from local file
    upgrade package git_svn-2.15.1-2 to 2.16.2-1 from local file
    upgrade package git_email-2.15.1-2 to 2.16.2-1 from local file
    upgrade package git_daemon-2.15.1-2 to 2.16.2-1 from local file
    upgrade package git_cvs-2.15.1-2 to 2.16.2-1 from local file
    upgrade package git_arch-2.15.1-2 to 2.16.2-1 from local file
Continue? [yes/no] (yes) : [system] Applying changes ...
[system] Changes applied. Old activation state backed up in "state_2018-03-09_13:19:39"
[system] Cleaning up ...
[system] Done.
```

`git add -i` works fine with perl 5.18.3:
```
*** Commands ***
  1: status	  2: update	  3: revert	  4: add untracked
  5: patch	  6: diff	  7: quit	  8: help
What now> 
```

**Comment by waddlesplash** [comment]
Unless anyone has a fix, yes, we should downgrade perl.

**Comment by fbrosson** [comment]
I got not fix. (Actually, I did not investigate, but I don't think it would be an easy task.)
So, shall I mark perl 5.26.2 as untested?

After that, I'll see if we need to bump the `REVISION` for perl 5.18.3. But I think it won't be necessary.

Once the builders have finished building perl 5.18.3 I'll bump `REVISION` for git 2.16.2.

Is everyone OK with this?

**Comment by waddlesplash** [comment]
Sounds fine.

**Comment by pulkomandy** [comment]
On Fri, Mar 09, 2018 at 11:08:12AM -0800, fbrosson wrote:
> I got not fix. (Actually, I did not investigate, but I don't think it would be an easy task.)
> So, shall I mark perl 5.26.2 as untested?
> 
> After that, I'll see if we need to bump the `REVISION` for perl 5.18.3. But I think it won't be necessary.
> 
> Once the builders have finished building perl 5.18.3 I'll bump `REVISION` for git 2.16.2.
> 
> Is everyone OK with this?

Not quite: it will break other packages!

The problem is (likely) that perm modules are stored in a
version-specific path (in this case, /system/lib/perl5/vendor_perl/5.18.2/Git/ )

Some Perl modules have already been rebuilt against 5.26 to get the
correct paths. But Git was not.

If you downgrade Perl, you will "fix" git, but you will also break
anything that was already rebuilt against the new Perl.

The solution is to make sure Git packages are built against the new
version of Perl so that the files ends up in the right place.

And we should probably make the Git package depend on a specific version
of perl (ideally, auto-populated at build time depending on what it was
built against).

**Comment by fbrosson** [comment]
> The solution is to make sure Git packages are built against the new version of Perl so that the files ends up in the right place.

In the git packages I have built against perl 5.26.1 the `Git.pm` is in `/boot/system/lib/perl5/vendor_perl/5.26/Git.pm`

But git is looking for `Git.pm` in `/boot/system/lib/perl5/vendor_perl/5.26.1/` (among several other locations), instead of `/boot/system/lib/perl5/vendor_perl/5.26/`.

**Comment by alaviss** [comment]
> But git is looking for Git.pm in `/boot/system/lib/perl5/vendor_perl/5.26.1/` (among several other locations), instead of `/boot/system/lib/perl5/vendor_perl/5.26/`.

This appears to be the culprit https://github.com/haikuports/haikuports/commit/39e5c6736760a81a930daefc01688caf75c8fddf.

However, we're not the only distribution that do so. Some debugging should be done to figure out why git was looking for perl modules there.

**Comment by fbrosson** [comment]
* f0dbf12da86ace6ca2f46da46238e3bd553560a9:  perl: use the shortVersion for site and vendor. (Thanks @korli)
* 0eb8b5970d6aba0b889be789ca69fe64d67ea63f: git: rebuild to fix "git add --interactive" after f0dbf12da86ace6ca2f46da46238e3bd553560a9

**Comment by fbrosson** [comment]
Shame on me. I just realized that it was not necessary to rebuild git!!!
